### ISSUE TITLE
[IMPROVEMENT] Prevent PVC provisioning if there is insufficent storage

### ISSUE THREAD

**Issue by PhanLe1010** [longhorn/longhorn#3529]
On the longhorn side we should prevent the volume creation, if there is no node capable of hosting it during the api volume creation call. We could return the Out of Range error, for this case during the CreateVolume csi call.
This evaluation should be done by the backend api creation call.

ref: https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume-errors

This issue is related to https://github.com/harvester/harvester/issues/1695

**Comment by PhanLe1010** [comment]
Related comment: https://github.com/longhorn/longhorn/issues/3626#issuecomment-1043206989

**Comment by derekbit** [comment]
`ReplicaScheduler` checks each node disks and finds the `preferredDisks` in https://github.com/longhorn/longhorn-manager/blob/master/scheduler/replica_scheduler.go#L246.

We can pass the all the disks' status to the upper caller.
But, the question is that a PVC creation failure might be due to multiple errors simultaneously, e.g. some disks have no sufficient storage and some disks' tags are not valid. In the case, how to represent the failure in Longhorn `volume.status.Conditions["scheduable"]`?

@PhanLe1010 @shuo-wu Do you have any idea?

cc. @innobead

**Comment by shuo-wu** [comment]
I don't think we need to worry too much about the failure message. Returning messages like `there is no available disk for the scheduling` would be fine. For this issue, the question is, how we can know if there is a disk that can fit a replica before creating the volume CR? CreationVolume API erroring out means the volume CR won't be created and we cannot rely on the volume-controller -> replica-scheduler to handle it.

**Comment by derekbit** [comment]
>  IMO, for this issue, the most important question is, how we can know if there is a disk that can fit a replica before creating the volume CR? 

@shuo-wu Yeah, I also think we cannot know the disks are invalid before creating the volume CR. Thus, I'm just going to set the appropriate error message in the volume condition. Currently, the condition message is empty.
Do you think it is okay?

**Comment by w13915984028** [comment]
From my view of point, the async (reverse) updating of the status of the PVC,  with a human understanding summary message, is acceptable now.  It is described in https://github.com/longhorn/longhorn/issues/3734.

Given the fact that the K8S system is loose-coupling and distributed, prevent the PVC provisioning in a real time manner based on storage, looks not easy.  

If LH (will) support PVC over-provisioning, it will be almost not practical to do so.

**Comment by shuo-wu** [comment]
You mean the Longhorn volume CR will be eventually created but an error will be replied to the VolumeCreation API call? This breaks one simple convention: If the creation API errors out, the object/resource is not actually created.

**Comment by derekbit** [comment]
No, the volumeCreation call is successful as current design. Only set error to the the volume CR condition.

**Comment by derekbit** [comment]
The LH differs from other storage projects in its design. The volume creation and replica scheduling are asynchronous. In LH, the volume CR is successfully created by the *CreateVolume* csi call, but the *ScheduleReplica* call may fail due to node or disk problems.

I check the error from `ScheduleReplica`, the errors can be classified into 
```
const (
	SchedulerErrorInsufficientStorage  = "insufficient storage"
	SchedulerErrorDiskNotFound         = "disk not found"
	SchedulerErrorSettingIOError       = "setting io error"
	SchedulerErrorsTagsNotFullfilled   = "tags not fullfilled"
	SchedulerErrorsNodeNotFound        = "node not found"
	SchedulerErrorsEngineImageNotReady = "engine image not ready"
)
```

Currently, the `Condition["Schedule"].message` is empty after the scheduling failure. We can update the it with a verbose error message is the combination of the errors, e.g. `insufficient storage` or `insufficient storage, disk not found`. Upper callers can then be aware of errors.

As for the reverse update of the PV/PVC, after discussion of @shuo-wu, in our opinion, the reverse update in LH is a bit of weird and seems not the responsibility of LH.

@PhanLe1010 @joshimoo @innobead WDYT?

cc @shuo-wu

**Comment by derekbit** [comment]
> For a workaround, Harvester tried a solution in bellow issue and PR. The main drawback is that Harvester will have full set of LH volumes, and they are frequently updated, this will increase CPU and memory usage. And Harvester deeply couples with LH implementation. 

But for the CPU usage caused by the periodical query to LH volume CR status from Harvester side is inevitable without reverse update of PV/PVC on LH side.

Another question is that although LH updates the verbose error message to PV/PVC annotation

**Comment by w13915984028** [comment]
Indeed, the reverse update of PV/PVC is not LH's responsibility.

Lets have a recall of WHY and HOW.

Harvester tried to solve via below PR, subscribes all LH volume resources in Harvester POD and deduces status by hard-coding.

The most important drawback is, the LH volume CR will cause Harvester POD has a significant increasing  CPU and memory resources, and LH volume will change all the time (every 5 seconds?).


https://github.com/harvester/harvester/pull/1923
![image](https://user-images.githubusercontent.com/31133476/164628059-05980f0b-fcd7-419a-bd1d-2ac1268c5557.png)

Deduce desired status.
![image](https://user-images.githubusercontent.com/31133476/164628189-323fd68c-72ad-4d86-8d85-6493a4fba5be.png)



After discussion with @guangbochen, @innobead , it is agreed that LH will have a solution to update to PVC/PV directly. Harvester will get such info from PVC/PV.
https://github.com/harvester/harvester/pull/1923#issuecomment-1064136206
for
https://github.com/longhorn/longhorn/issues/3734

When both LH and Harvester do a bit more, we will solve a focal point from customers' perspective.


In furure, CSI may add new interfaces to let driver layer report status, or have APIs for PVC/PV to query.

**Comment by derekbit** [comment]
Thanks for @w13915984028 's explanation.

Before the new CSI's new interface, looks LH needs to update the PV annotation for the pain point temporarily.

BTW, I also checked other projects, and there are no similar issue reported and maybe due to the LH's async design for volume/replica.

**Comment by shuo-wu** [comment]
The Longhorn volume frequent update is probably related to `volume.Status.ActualSize`, which will be synced with the corresponding engine every 5 seconds.

**Comment by yangchiu** [comment]
verified passed on v1.3.0-preview1 by:

1.
1.a. Disable all disks and on all nodes
2.b. Create a volume on UI
3.c. Create PV/PVC for the volume
4.d. Check the PV annotation "longhorn.io/volume-scheduling-error", the error message is "disks are not available"

2
2.a. Create a volume on UI. The request size exceeds the disk available size.
2.b. Create PV/PVC for the volume
2.c. Check the PV annotation "longhorn.io/volume-scheduling-error", the error message is "insufficient storage"

3. Repeat the 1 and 2, but create the volume via PVC

**Comment by futuretea** [comment]
@guangbochen @PhanLe1010 
Test 3 cases in harvester v1.1.0-rc (with longhorn v1.3.1), It seems that the result is not the expected

case 1:
- create a PVC with size > 8589934591 GiB

The result is the PVC was in pending state, no PV and longhorn volume 

<details>
<summary>yamls</summary>

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    volume.beta.kubernetes.io/storage-provisioner: driver.longhorn.io
    volume.kubernetes.io/storage-provisioner: driver.longhorn.io
  creationTimestamp: "2022-09-30T03:48:50Z"
  finalizers:
  - kubernetes.io/pvc-protection
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:spec:
        f:accessModes: {}
        f:resources:
          f:requests:
            .: {}
            f:storage: {}
        f:storageClassName: {}
        f:volumeMode: {}
    manager: harvester
    operation: Update
    time: "2022-09-30T03:48:50Z"
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:volume.beta.kubernetes.io/storage-provisioner: {}
          f:volume.kubernetes.io/storage-provisioner: {}
    manager: kube-controller-manager
    operation: Update
    time: "2022-09-30T03:48:50Z"
  name: aaaaa
  namespace: default
  resourceVersion: "5013896"
  uid: 463d249f-e54f-4d64-9223-ff19d95c9b99
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: "9223372036854775807"
  storageClassName: harvester-longhorn
  volumeMode: Block
status:
  phase: Pending
```
</details>
case 2:

- Create a PVC with size  <= 8589934591 GiB, but greater than the actual available space size.

The result is PV and longhorn volume will be created, but the annotation `longhorn.io/volume-scheduling-error` value in PV is empty

<details>
<summary>yamls</summary>

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.beta.kubernetes.io/storage-provisioner: driver.longhorn.io
    volume.kubernetes.io/storage-provisioner: driver.longhorn.io
  creationTimestamp: "2022-09-30T09:21:39Z"
  finalizers:
  - kubernetes.io/pvc-protection
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:spec:
        f:accessModes: {}
        f:resources:
          f:requests:
            .: {}
            f:storage: {}
        f:storageClassName: {}
        f:volumeMode: {}
    manager: harvester
    operation: Update
    time: "2022-09-30T09:21:39Z"
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:pv.kubernetes.io/bind-completed: {}
          f:pv.kubernetes.io/bound-by-controller: {}
          f:volume.beta.kubernetes.io/storage-provisioner: {}
          f:volume.kubernetes.io/storage-provisioner: {}
      f:spec:
        f:volumeName: {}
    manager: kube-controller-manager
    operation: Update
    time: "2022-09-30T09:21:41Z"
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:accessModes: {}
        f:capacity:
          .: {}
          f:storage: {}
        f:phase: {}
    manager: kube-controller-manager
    operation: Update
    subresource: status
    time: "2022-09-30T09:21:41Z"
  name: bbbbbbbbb
  namespace: default
  resourceVersion: "5276144"
  uid: 214c71f7-9136-4839-8a81-a00c8562b09d
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 8589934591Gi
  storageClassName: harvester-longhorn
  volumeMode: Block
  volumeName: pvc-214c71f7-9136-4839-8a81-a00c8562b09d
status:
  accessModes:
  - ReadWriteMany
  capacity:
    storage: 8589934591Gi
  phase: Bound
```
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    longhorn.io/volume-scheduling-error: ""
    pv.kubernetes.io/provisioned-by: driver.longhorn.io
  creationTimestamp: "2022-09-30T09:21:41Z"
  finalizers:
  - kubernetes.io/pv-protection
  name: pvc-214c71f7-9136-4839-8a81-a00c8562b09d
  resourceVersion: "5276143"
  uid: 61549ca5-1125-4856-bc5d-31564a61c084
spec:
  accessModes:
  - ReadWriteMany
  capacity:
    storage: 8589934591Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: bbbbbbbbb
    namespace: default
    resourceVersion: "5276083"
    uid: 214c71f7-9136-4839-8a81-a00c8562b09d
  csi:
    driver: driver.longhorn.io
    volumeAttributes:
      baseImage: ""
      fromBackup: ""
      migratable: "true"
      numberOfReplicas: "3"
      share: "true"
      staleReplicaTimeout: "30"
      storage.kubernetes.io/csiProvisionerIdentity: 1664508845257-8081-driver.longhorn.io
    volumeHandle: pvc-214c71f7-9136-4839-8a81-a00c8562b09d
  persistentVolumeReclaimPolicy: Delete
  storageClassName: harvester-longhorn
  volumeMode: Block
status:
  phase: Bound
```
```yaml
apiVersion: longhorn.io/v1beta2
kind: Volume
metadata:
  creationTimestamp: "2022-09-30T09:21:39Z"
  finalizers:
  - longhorn.io
  generation: 1
  labels:
    longhornvolume: pvc-214c71f7-9136-4839-8a81-a00c8562b09d
    recurring-job-group.longhorn.io/default: enabled
  name: pvc-214c71f7-9136-4839-8a81-a00c8562b09d
  namespace: longhorn-system
  resourceVersion: "5276145"
  uid: 0b197cfd-402e-42ba-8e4e-95c455fde953
spec:
  Standby: false
  accessMode: rwx
  backingImage: ""
  baseImage: ""
  dataLocality: disabled
  dataSource: ""
  disableFrontend: false
  diskSelector: []
  encrypted: false
  engineImage: longhornio/longhorn-engine:v1.3.1
  fromBackup: ""
  frontend: blockdev
  lastAttachedBy: ""
  migratable: true
  migrationNodeID: ""
  nodeID: ""
  nodeSelector: []
  numberOfReplicas: 3
  recurringJobs: []
  replicaAutoBalance: ignored
  revisionCounterDisabled: false
  size: "9223372035781033984"
  staleReplicaTimeout: 30
status:
  actualSize: 0
  cloneStatus:
    snapshot: ""
    sourceVolume: ""
    state: ""
  conditions:
  - lastProbeTime: ""
    lastTransitionTime: "2022-09-30T09:21:39Z"
    message: ""
    reason: ""
    status: "False"
    type: toomanysnapshots
  - lastProbeTime: ""
    lastTransitionTime: "2022-09-30T09:21:39Z"
    message: Reset schedulable due to allow volume creation with degraded availability
    reason: ""
    status: "True"
    type: scheduled
  - lastProbeTime: ""
    lastTransitionTime: "2022-09-30T09:21:39Z"
    message: ""
    reason: ""
    status: "False"
    type: restore
  currentImage: longhornio/longhorn-engine:v1.3.1
  currentNodeID: ""
  expansionRequired: false
  frontendDisabled: false
  isStandby: false
  kubernetesStatus:
    lastPVCRefAt: ""
    lastPodRefAt: ""
    namespace: default
    pvName: pvc-214c71f7-9136-4839-8a81-a00c8562b09d
    pvStatus: Bound
    pvcName: bbbbbbbbb
    workloadsStatus: null
  lastBackup: ""
  lastBackupAt: ""
  lastDegradedAt: ""
  ownerID: harv-11rc1
  pendingNodeID: ""
  remountRequestedAt: ""
  restoreInitiated: false
  restoreRequired: false
  robustness: unknown
  shareEndpoint: ""
  shareState: ""
  state: detached
```
</details>

case 3:

- Create a PVC with size < the actual available space size，
- Resize the PVC to a not schedulable size

The result is PV and longhorn volume will be created, but the annotation `longhorn.io/volume-scheduling-error` value in PV is empty
<details>
<summary>yamls</summary>

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.beta.kubernetes.io/storage-provisioner: driver.longhorn.io
    volume.kubernetes.io/storage-provisioner: driver.longhorn.io
  creationTimestamp: "2022-09-30T09:23:06Z"
  finalizers:
  - kubernetes.io/pvc-protection
  name: cccccccccc
  namespace: default
  resourceVersion: "5278110"
  uid: 82395853-35de-42c3-8394-45997518cc12
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 100000Gi
  storageClassName: harvester-longhorn
  volumeMode: Block
  volumeName: pvc-82395853-35de-42c3-8394-45997518cc12
status:
  accessModes:
  - ReadWriteMany
  capacity:
    storage: 1Gi
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2022-09-30T09:24:07Z"
    status: "True"
    type: Resizing
  phase: Bound
```
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    longhorn.io/volume-scheduling-error: ""
    pv.kubernetes.io/provisioned-by: driver.longhorn.io
  creationTimestamp: "2022-09-30T09:23:08Z"
  finalizers:
  - kubernetes.io/pv-protection
  name: pvc-82395853-35de-42c3-8394-45997518cc12
  resourceVersion: "5277338"
  uid: 4d6e93bb-bade-410e-999c-43582fd8444d
spec:
  accessModes:
  - ReadWriteMany
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: cccccccccc
    namespace: default
    resourceVersion: "5277280"
    uid: 82395853-35de-42c3-8394-45997518cc12
  csi:
    driver: driver.longhorn.io
    volumeAttributes:
      baseImage: ""
      fromBackup: ""
      migratable: "true"
      numberOfReplicas: "3"
      share: "true"
      staleReplicaTimeout: "30"
      storage.kubernetes.io/csiProvisionerIdentity: 1664508845257-8081-driver.longhorn.io
    volumeHandle: pvc-82395853-35de-42c3-8394-45997518cc12
  persistentVolumeReclaimPolicy: Delete
  storageClassName: harvester-longhorn
  volumeMode: Block
status:
  phase: Bound
```
```yaml
apiVersion: longhorn.io/v1beta2
kind: Volume
metadata:
  creationTimestamp: "2022-09-30T09:23:06Z"
  finalizers:
  - longhorn.io
  generation: 2
  labels:
    longhornvolume: pvc-82395853-35de-42c3-8394-45997518cc12
    recurring-job-group.longhorn.io/default: enabled
  name: pvc-82395853-35de-42c3-8394-45997518cc12
  namespace: longhorn-system
  resourceVersion: "5278265"
  uid: f000c91b-facb-4097-bd38-4ce320359851
spec:
  Standby: false
  accessMode: rwx
  backingImage: ""
  baseImage: ""
  dataLocality: disabled
  dataSource: ""
  disableFrontend: true
  diskSelector: []
  encrypted: false
  engineImage: longhornio/longhorn-engine:v1.3.1
  fromBackup: ""
  frontend: blockdev
  lastAttachedBy: ""
  migratable: true
  migrationNodeID: ""
  nodeID: ""
  nodeSelector: []
  numberOfReplicas: 3
  recurringJobs: []
  replicaAutoBalance: ignored
  revisionCounterDisabled: false
  size: "107374182400000"
  staleReplicaTimeout: 30
status:
  actualSize: 0
  cloneStatus:
    snapshot: ""
    sourceVolume: ""
    state: ""
  conditions:
  - lastProbeTime: ""
    lastTransitionTime: "2022-09-30T09:23:06Z"
    message: ""
    reason: ""
    status: "False"
    type: toomanysnapshots
  - lastProbeTime: ""
    lastTransitionTime: "2022-09-30T09:24:11Z"
    message: ""
    reason: ""
    status: "True"
    type: scheduled
  - lastProbeTime: ""
    lastTransitionTime: "2022-09-30T09:23:06Z"
    message: ""
    reason: ""
    status: "False"
    type: restore
  currentImage: longhornio/longhorn-engine:v1.3.1
  currentNodeID: ""
  expansionRequired: true
  frontendDisabled: true
  isStandby: false
  kubernetesStatus:
    lastPVCRefAt: ""
    lastPodRefAt: ""
    namespace: default
    pvName: pvc-82395853-35de-42c3-8394-45997518cc12
    pvStatus: Bound
    pvcName: cccccccccc
    workloadsStatus: null
  lastBackup: ""
  lastBackupAt: ""
  lastDegradedAt: ""
  ownerID: harv-11rc1
  pendingNodeID: ""
  remountRequestedAt: ""
  restoreInitiated: false
  restoreRequired: false
  robustness: faulted
  shareEndpoint: ""
  shareState: ""
  state: detached
```
</details>

**Comment by innobead** [comment]
@derekbit Could you help check the above possible regression?

**Comment by derekbit** [comment]
The request is rejected by the admission webhook before scheduling the volume.
So `longhorn.io/volume-scheduling-error` won’t be set.

**Comment by derekbit** [comment]
- Case 1
  - If you create a PVC with size larger than 8589934591 GiB, an int64  (volume.size) overflow happens. Then, the csi driver will receive an zero-sized volume creation request. The zero-sized volume creation request should be rejected in the mutating webhook. However, the overflow issue is not 
- Case 2
  - Works well in RWO volumes
  - But the `longhorn.io/volume-scheduling-error` is not set corectly in RWX volumes.
- Case 3
  - Same as Case 2

**Comment by innobead** [comment]
Because this issue has been closed for 1.2.5 and 1.3.0. @derekbit Will create another issue for 1.4.0 and related patch fixes.

**Comment by derekbit** [comment]
Track in https://github.com/longhorn/longhorn/issues/4654 instead.
